src/Preprocessor.zig

```diff
@@ -510,14 +513,400 @@ fn skipToNl(tokenizer: *Tokenizer) void {
+    _ = tokenizer;
+    var moving_end_idx = end_idx;
+    // rescan loop
+    var do_rescan = true;
+    while (do_rescan) {
+        do_rescan = false;
+        //std.debug.print("Scanning ", .{});
+        //try pp.debugTokenBuf(buf.*);
+        // expansion loop
+        var idx: usize = start_idx;
+        while (idx < moving_end_idx) {
+            const macro_entry = pp.defines.getPtr(pp.expandedSlice(buf.items[idx]));
+            if (macro_entry != null and shouldExpand(buf.items[idx], macro_entry.?)) {
+                switch (macro_entry.?.*) {
+                    .empty => {
+                        idx += 1;
+                    },
+                    .simple => |simple_macro| {
+                        //std.debug.print("Expanding {s}\n", .{pp.expandedSlice(buf.items[idx])});
+                        const res = try pp.expandObjMacro(&simple_macro);
+                        defer res.deinit();
+                        var expansion_loc = simple_macro.loc;
+                        expansion_loc.next = buf.items[idx].loc.next;
+                        for (res.items) |*tok| {
+                            if (buf.items[idx].loc.next) |ln| {
+                                try pp.markExpandedFrom(tok, ln.*);
+                            }
+                            try pp.markExpandedFrom(tok, simple_macro.loc);
+                        }
+
+                        try buf.replaceRange(idx, 1, res.items);
+                        idx += res.items.len;
+                        moving_end_idx += (res.items.len - 1);
+                        do_rescan = true;
+                    },
+                    .func => |func_macro| {
+                        var macro_scan_idx = idx;
+                        if (try pp.collectMacroFuncArguments(tokenizer, buf, &macro_scan_idx, &moving_end_idx, extend_buf)) |args| {
+                            // TODO: bug?
+                            //defer for (args.items) |item| {
+                            //    pp.comp.gpa.free(item);
+                            //};
+                            //defer args.deinit();
+                            var args_count = @intCast(u32, args.items.len);
+                            // if the macro has zero arguments g() args_count is still 1
+                            if (args_count == 1 and func_macro.params.len == 0) args_count = 0;
+
+                            // Validate argument count.
+                            const extra = Diagnostics.Message.Extra{ .arguments = .{ .expected = @intCast(u32, func_macro.params.len), .actual = args_count } };
+                            var validated = true;
+                            if (func_macro.var_args and args_count < func_macro.params.len) {
+                                try pp.comp.diag.add(.{ .tag = .expected_at_least_arguments, .loc = buf.items[idx].loc, .extra = extra });
+                                validated = false;
```

Comment:
```suggestion
                                idx += 1;
                                continue;
```
We're there some issues here earlier?

Reply:
I'm not getting what you are suggesting, if validated = true the next if would be entered, which must expand the macro func, but there are not enough arguments to do so.. am I missing something?

Reply:
The `validated` flag is complicating the control flow. In both cases if `validated == false` the equivalent of `idx +=1; continue;` will be executed. Removing it simplifies the control flow and reduces the nesting depth.

I originally didn't notice the `do_rescan` at the end of the loop but that should now be accounted for by an earlier comment.